src/persist-client/src/project.rs

```diff
@@ -0,0 +1,181 @@
+    /// aggressive [Self::try_optimize_ignored_data_fetch] optimization and it
+    /// corresponds to a common query shape: `SELECT count(*)`.
+    ///
+    /// This error bit is certainly a bit of an abstraction breakage in the
+    /// "persist is independent of mz" story, but it should go away when we
+    /// implement full projection pushdown.
+    IgnoreAllNonErr {
+        /// The name of the top-level error column.
+        err_col_name: &'static str,
+        /// The `Codec` encoded key corresponding to ignored data.
+        key_bytes: Vec<u8>,
+        /// The `Codec` encoded val corresponding to ignored data.
+        val_bytes: Vec<u8>,
+    },
+}
+
+impl Default for ProjectionPushdown {
+    fn default() -> Self {
+        ProjectionPushdown::FetchAll
+    }
+}
+
+impl ProjectionPushdown {
+    /// If relevant, applies the [Self::IgnoreAllNonErr] projection to a part
+    /// about to be fetched.
+    ///
+    /// If a part contains data from entirely before a snapshot `as_of`, and the
+    /// pushed-down MFP projects to an empty list of columns, and we can prove
+    /// that the part is error free, then we can use the diff sum from stats
+    /// instead of loading the data. In this case, we return a `Some` with a
+    /// replacement [BatchPart]. In all other cases, a None.
+    ///
+    /// - Summing the diffs in a part is equivalent to projecting the row in
+    ///   each tuple to an empty row and then consolidating. (Which is pretty
+    ///   much how `select count(*)` queries get compiled today.)
+    /// - If the as-of timestamp falls in the middle of a part, we can just
+    ///   fetch and process the part as normal. The optimization can still
+    ///   provide a speedup for other parts. TODO: We could improve this by
+    ///   keeping track in metadata of the largest timestamp(s) in a hollow
+    ///   part.
+    pub(crate) fn try_optimize_ignored_data_fetch<T: Timestamp + Codec64>(
+        &self,
+        cfg: &ConfigSet,
+        metrics: &Metrics,
+        as_of: &Antichain<T>,
+        desc: &Description<T>,
+        part: &BatchPart<T>,
+    ) -> Option<BatchPart<T>> {
+        if !OPTIMIZE_IGNORED_DATA_FETCH.get(cfg) {
+            return None;
+        }
+        let (err_col_name, key_bytes, val_bytes) = match self {
+            ProjectionPushdown::FetchAll => return None,
+            ProjectionPushdown::IgnoreAllNonErr {
+                err_col_name,
+                key_bytes,
+                val_bytes,
+            } => (*err_col_name, key_bytes.as_slice(), val_bytes.as_slice()),
+        };
+        let (diffs_sum, stats) = match &part {
+            BatchPart::Hollow(x) => (x.diffs_sum, x.stats.as_ref()),
+            BatchPart::Inline { .. } => return None,
+        };
+        debug!(
+            "try_optimize_ignored_data_fetch diffs_sum={:?} as_of={:?} lower={:?} upper={:?}",
+            diffs_sum.map(i64::decode),
```

Comment:
This will look a bit weird if `diffs_sum` is not i64, but that seems fine for a debug output...

Reply:
There's a couple places where we do something like this for debug output. The others have a comment but I got lazy here. Added one here too!